source/MaterialXGenOgsXml/OgsFragment.cpp

```diff
@@ -0,0 +1,353 @@
+#include "OgsFragment.h"
+
+#include <MaterialXFormat/XmlIo.h>
+#include <MaterialXGenShader/Util.h>
+#include <MaterialXGenShader/DefaultColorManagementSystem.h>
+#include <MaterialXGenOgsXml/GlslFragmentGenerator.h>
+#include <MaterialXGenShader/GenContext.h>
+#include <MaterialXGenOgsXml/OgsXmlGenerator.h>
+
+#ifdef MATERIALX_BUILD_CROSS
+#include <MaterialXCross/Cross.h>
+#endif
+
+#include <iostream>
+
+namespace MaterialXMaya
+{
+namespace
+{
+// The base class for classes wrapping GLSL fragment generators for use during
+// OgsFragment construction.
+class GlslGeneratorWrapperBase
+{
+    GlslGeneratorWrapperBase() = delete;
+
+  protected:
+    explicit GlslGeneratorWrapperBase(mx::ElementPtr element)
+        : _element(element)
+    {
+        if (!_element)
+            throw mx::Exception("No element specified");
+
+        if (element->asA<mx::ShaderRef>())
+        {
+            _isSurface = true;
+        }
+        else if (!element->asA<mx::Output>())
+        {
+            throw mx::Exception("Invalid element to create fragment for " + element->getName());
+        }
+    }
+
+  protected:
+    void setCommonOptions(mx::GenOptions& genOptions, const mx::ShaderGenerator& generator)
+    {
+        // Use FIS environment lookup for surface shader generation but
+        // disable for texture nodes to avoid additional unneeded XML parameter
+        // generation.    
+        genOptions.hwSpecularEnvironmentMethod =
+            _isSurface ? mx::SPECULAR_ENVIRONMENT_FIS : mx::SPECULAR_ENVIRONMENT_NONE;
+
+        // Set to use no direct lighting
+        genOptions.hwMaxActiveLightSources = 0;
+
+        // Maya images require a texture coordinates to be flipped in V.
+        genOptions.fileTextureVerticalFlip = true;
+        genOptions.hwTransparency = mx::isTransparentSurface(_element, generator);
+
+        // Maya viewport uses texture atlas for tile image so enabled
+        // texture coordinate transform to go from original UDIM range to
+        // normalized 0..1 range.
+        genOptions.hwNormalizeUdimTexCoords = true;
+    }
+
+    mx::ElementPtr _element;
+
+private:
+    bool _isSurface = false;
+};
+
+// Knows how to create a temporary local GLSL fragment generator to generate
+// GLSL fragment code during OgsFragment construction.
+//
+class LocalGlslGeneratorWrapper
```

Comment:
Minor usage naming: I'm not sure what "local" means. Is this a "internal" to the "external", or is this a "private" to "public" interface difference?.

Reply:
Well, the generator and the context are literally local variables in this case, destroyed right after usage.